Broken dereferencing when using autogenerated str() fns

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.1.30 50a3009.d60f148
**OS:** windows, Microsoft Windows 10 Pro v19042 64-bit

**What did you do?**
Tried compiling
```v
struct Foo {
	bar int 
}

fn (f &Foo) str() string {
	return '${f.bar}'
}

struct Bar  {
	foo &Foo
}

fn main() {
	foo := Foo{}
	bar := Bar { &foo }
	println(bar)
}
```
**What did you expect to see?**
✨A successful compilation✨
 
**What did you see instead?**
```
PS C:\Users\hkg90\Projects\advent-of-code-2020> v -cg .\bad_dereference.v
recompilation with tcc failed; retrying with cc ...
C:\Users\hkg90\AppData\Local\Temp\v\bad_dereference.17764375660490729857.tmp.c: In function 'indent_main__Bar_str':
C:\Users\hkg90\AppData\Local\Temp\v\bad_dereference.17764375660490729857.tmp.c:1481:57: error: incompatible type for argument 1 of 'main__Foo_str' 
   indents, isnil(it.foo) ? _SLIT("nil") : main__Foo_str(*it.foo),
                                                         ^~~~~~~
C:\Users\hkg90\AppData\Local\Temp\v\bad_dereference.17764375660490729857.tmp.c:1448:33: note: expected 'main__Foo *' {aka 'struct main__Foo *'} but argument is of type 'main__Foo' {aka 'struct main__Foo'}
 string main__Foo_str(main__Foo* f);
```

___
`Bar.foo` is a ref, `main__Foo_str` expected a ref but codegen dereferenced the `foo`
```c
	return _STR("Bar{\n"		"%.*s\000    foo: &%.*s\000\n"
		"%.*s\000}", 4,
		indents, isnil(it.foo) ? _SLIT("nil") : main__Foo_str(*it.foo),
		//													 ~^~
		indents);
);
```

The faulty line of code (I think) is located at https://github.com/vlang/v/blob/master/vlib/v/gen/auto_str_methods.v#L436

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -469,7 +469,7 @@ fn (mut g Gen) gen_str_for_struct(info table.Struct, styp string, str_fn_name st
 	if clean_struct_v_type_name.contains('_T_') {
 		// TODO: this is a bit hacky. styp shouldn't be even parsed with _T_
 		// use something different than g.typ for styp
-		clean_struct_v_type_name = 
+		clean_struct_v_type_name =
 			clean_struct_v_type_name.replace('_T_', '<').replace('_', ', ').replace('Array', 'array') +
 			'>'
 	}
@@ -530,12 +530,12 @@ fn (mut g Gen) gen_str_for_struct(info table.Struct, styp string, str_fn_name st
 }
 
 fn struct_auto_str_func(sym table.TypeSymbol, field_type table.Type, fn_name string, field_name string) string {
-	has_custom_str := sym.has_method('str')
+	has_custom_str, expects_ptr, _ := sym.str_method_info()
 	if sym.kind in [.enum_, .interface_] {
 		return '${fn_name}(it.${c_name(field_name)})'
 	} else if sym.kind == .struct_ {
 		mut obj := 'it.${c_name(field_name)}'
-		if field_type.is_ptr() {
+		if field_type.is_ptr() && !expects_ptr {
 			obj = '*$obj'
 		}
 		if has_custom_str {
@@ -553,7 +553,7 @@ fn struct_auto_str_func(sym table.TypeSymbol, field_type table.Type, fn_name str
 		mut method_str := 'it.${c_name(field_name)}'
 		if sym.kind == .bool {
 			method_str += ' ? _SLIT("true") : _SLIT("false")'
-		} else if (field_type.is_int() || field_type.is_float()) && field_type.is_ptr() {
+		} else if (field_type.is_int() || field_type.is_float()) && field_type.is_ptr() && !expects_ptr {
 			// ptr int can be "nil", so this needs to be castet to a string
 			fmt := if sym.kind in [.f32, .f64] {
 				'%g\\000'

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -469,7 +469,7 @@ fn (mut g Gen) gen_str_for_struct(info table.Struct, styp string, str_fn_name st
 	if clean_struct_v_type_name.contains('_T_') {
 		// TODO: this is a bit hacky. styp shouldn't be even parsed with _T_
 		// use something different than g.typ for styp
-		clean_struct_v_type_name =
+		clean_struct_v_type_name = 
 			clean_struct_v_type_name.replace('_T_', '<').replace('_', ', ').replace('Array', 'array') +
 			'>'
 	}
@@ -553,7 +553,8 @@ fn struct_auto_str_func(sym table.TypeSymbol, field_type table.Type, fn_name str
 		mut method_str := 'it.${c_name(field_name)}'
 		if sym.kind == .bool {
 			method_str += ' ? _SLIT("true") : _SLIT("false")'
-		} else if (field_type.is_int() || field_type.is_float()) && field_type.is_ptr() && !expects_ptr {
+		} else if (field_type.is_int() || field_type.is_float()) && field_type.is_ptr()
+			&& !expects_ptr {
 			// ptr int can be "nil", so this needs to be castet to a string
 			fmt := if sym.kind in [.f32, .f64] {
 				'%g\\000'

--- a/vlib/v/tests/string_interpolation_custom_str_test.v
+++ b/vlib/v/tests/string_interpolation_custom_str_test.v
@@ -1,0 +1,20 @@
+struct Foo {
+	bar int
+}
+
+fn (f &Foo) str() string {
+	return '${f.bar}'
+}
+
+struct Bar  {
+	foo &Foo
+}
+
+fn test_interpolation_with_custom_ref_str() {
+	foo := Foo{}
+	bar := Bar { &foo }
+	println(bar)
+	assert '$bar'.contains('Bar{')
+	assert '$bar'.contains('foo: &0')
+	assert '$bar'.contains('}')
+}

